homeassistant/components/fritzbox_callmonitor/sensor.py

```diff
@@ -133,21 +156,48 @@ def name(self):
         """Return the name of the sensor."""
         return self._name
 
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return ICON_PHONE
+
     @property
     def device_state_attributes(self):
         """Return the state attributes."""
+        if self._prefixes:
+            self._attributes[ATTR_PREFIXES] = self._prefixes
+        if self._phonebook_name:
+            self._attributes[ATTR_PHONEBOOK_NAME] = self._phonebook_name
+        if self._phonebook_url:
+            self._attributes[ATTR_PHONEBOOK_URL] = self._phonebook_url
         return self._attributes
 
+    @property
+    def device_info(self):
+        """Return device specific attributes."""
+        return {
+            "name": self._phonebook.fph.modelname,
+            "identifiers": {(DOMAIN, self._host, self._phonebook_id)},
+            "manufacturer": MANUFACTURER,
+            "model": self._phonebook.fph.modelname,
+            "sw_version": self._phonebook.fph.fc.system_version,
+        }
+
+    @property
+    def unique_id(self):
+        """Return the unique ID of the device."""
+        return f"{self._host}-{self._phonebook_id}"
+
     def number_to_name(self, number):
         """Return a name for a given phone number."""
-        if self.phonebook is None:
+        if self._phonebook is None:
             return "unknown"
-        return self.phonebook.get_name(number)
+        return self._phonebook.get_name(number)
 
     def update(self):
         """Update the phonebook if it is defined."""
-        if self.phonebook is not None:
-            self.phonebook.update_phonebook()
+        if self._phonebook is not None:
+            self._phonebook.update_phonebook()
 
 
 class FritzBoxCallMonitor:
```

Comment:
This whole class shoud be in the PyPI package.

Reply:
Completely agree that the socket connection should be part of the Pypi package, however this class is also neatly intertwined with the created sensor by setting its state and attributes. How could I achieve this by extracting the socket connection to the Pypi package?

Reply:
Create a class that is being mutated by the socket on which the FritzEntity can register a callback to be called when new data is available.

Reply:
Just so I understand this correctly: I should extract the entire `FritzBoxCallMonitor` into the Pypi package, remove the `sensor` attribute and add an attribute, the callback function, that should be called when new data is available?

Something like this?
```python
self._monitor = package.FritzBoxCallMonitor(
            host=self._host,
            port=self._port,
            callback_function=self.update_state_and_attributes,
        )

def update_state_and_attributes(self):
    data = self._monitor.data
    self._attributes = {...}
    self._state = data.state
```